vault/diagnose/diagnose_tls_test.go

```diff
@@ -0,0 +1,146 @@
+					TLSClientCAFile:               "./certs/rootca3.crt",
+					TLSMinVersion:                 "tls13",
+					TLSRequireAndVerifyClientCert: false,
+					TLSDisableClientCerts:         true,
+				},
+			},
+		},
+	}
+	conf := &vault.CoreConfig{
+		RawConfig: serverConf,
+	}
+	core := vault.TestCoreWithConfig(t, conf)
+	return core
+}
+
+func TestTLSValidCert(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./../../api/test-fixtures/keys/cert.pem",
+				TLSKeyFile:                    "./../../api/test-fixtures/keys/key.pem",
+				TLSClientCAFile:               "./../../api/test-fixtures/root/rootcacert.pem",
+				TLSMinVersion:                 "0",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err != nil {
+		t.Fail()
+	}
+}
+
+func TestTLSFakeCert(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./test-fixtures/fakecert.pem",
+				TLSKeyFile:                    "./../../api/test-fixtures/keys/key.pem",
+				TLSClientCAFile:               "./../../api/test-fixtures/root/rootcacert.pem",
+				TLSMinVersion:                 "0",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err == nil {
+		t.Errorf("TLS Config check on fake certificate should fail")
+	}
+	if err.Error() != "tls: failed to find any PEM data in certificate input" {
+		t.Errorf("Bad error message: %s", err.Error())
+	}
+}
+
+// TestTLSTrailingData uses a certificate from:
+// https://github.com/golang/go/issues/40545 that contains
+// an extra DER sequence, and makes sure a trailing data error
+// is returned.
+func TestTLSTrailingData(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./test-fixtures/trailingdatacert.pem",
+				TLSKeyFile:                    "./../../api/test-fixtures/keys/key.pem",
+				TLSClientCAFile:               "./../../api/test-fixtures/root/rootcacert.pem",
+				TLSMinVersion:                 "0",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err == nil {
+		t.Errorf("TLS Config check on fake certificate should fail")
+	}
+	if err.Error() != "asn1: syntax error: trailing data" {
```

Comment:
does this error string exist in some library, or must it be hard coded this way?

Reply:
It exists in a library.

Reply:
so is that string signifying trailing data not defined somewhere in the library then?

Reply:
It's not defined as a static string that I can import. It's defined like so: `err = asn1.SyntaxError{Msg: "trailing data"}` in pkcs7.go

Reply:
ah, bummer! so either way you'd still have to have some static string in there. alright!